### PR TITLE
feat: refactor object models to simplify offline storage

### PR DESCRIPTION
**Pull request by dmccartney**
This restructures the object model to be more offline-storage friendly.

Also included:
- fixed a bug with preKey signing
- simplified the interface to `conversationId`/`metadata`

### CODE REVIEW

File: lib/src/content/decoded.dart [hunk 1]
```diff
@@ -3,30 +3,78 @@ import 'package:xmtp_proto/xmtp_proto.dart' as xmtp;
...
+/// and want to list the messages sequentially.
+/// To support this you'll want to key by the [topic] and [sentAt] together.
+///  -> Tip: consider [topic]+[sentAt] as an index in the local database.
+///
+/// Storing the decoded [content] is difficult because it can be any type.
+/// Instead you'll want to store the [encoded] content because it can be
+/// reliably serialized. Later, you can decode it using [Client.decodeContent].
+///  -> Tip: store the encoded content using `encoded.writeToBuffer()`
+///
+/// See the example app for a demonstration of the overall approach.
```
Comment (dmccartney) [inline lib/src/content/decoded.dart:25]: here's the gist of my offline storage guidance

File: lib/xmtp.dart [hunk 2]
```diff
@@ -1,4 +1,7 @@
 library xmtp;
 
-// TODO
-// export ...
+export 'src/common/api.dart' show Api;
+export 'src/client.dart' show Client;
+export 'src/conversation/conversation.dart' show Conversation;
+export 'src/content/decoded.dart' show DecodedMessage;
+export 'package:xmtp_proto/xmtp_proto.dart';
```
Comment (dmccartney) [inline lib/xmtp.dart:7]: This is the first cut at defining what, exactly, is exposed from the SDK.

File: test/auth_test.dart [hunk 3]
```diff
@@ -25,6 +26,16 @@ void main() {
...
+    expect(
+      token.identityKey.recoverWalletSignerPublicKey().toEthereumAddress(),
+      alice.address,
+    );
+    // Note: there's a difference here between the contact and auth token.
+    //       The go backend expects auth tokens signed with `ecdsaCompact`.
+    //       The js-lib expects contacts signed with `walletEcdsaCompact`.
+    // TODO: teach both ^ to accept either.
+    // For now, this is what the backend expects inside the authToken.
+    expect(token.identityKey.signature.hasEcdsaCompact(), true);
```
Comment (dmccartney) [inline test/auth_test.dart:38]: This is the gist of the signature-shuffling I was referring to before.

File: test/client_test.dart [hunk 4]
```diff
@@ -91,7 +132,7 @@ void main() {
       for (var convo in conversations) {
         debugPrint('Conversation with ${convo.peer} ${convo.version}');
         debugPrint(' -> ${convo.topic}');
-        var messages = await convo.listMessages();
+        var messages = await client.listMessages(convo);
         for (var msg in messages) {
```
Comment (dmccartney) [inline test/client_test.dart:136]: This demonstrates the gist of the object model change.

Both `DecodedMessage` and especially `Conversation` are more like identifying records that can be serialized (fewer verbs, fewer dependencies). The actions (listing messages, sending messages) are done by the `Client` instead (with the `Conversation` as a parameter).

File: test/auth_test.dart (hunk 3 above)
Comment (dmccartney) [reply test/auth_test.dart:38]: ref https://github.com/xmtp/xmtp-node-go/pull/193

**Review by elisealix22 — APPROVED** [review]
I see all those docs coming together! Looks great. Just one suggestion to link over to the example if we can 🙏

File: lib/src/content/decoded.dart (hunk 1 above)
Comment (dmccartney) [reply lib/src/content/decoded.dart:25]: Absolutely will do. I'm about to put up the PR with the `example/` app inside. I'll make sure to update these to link directly into it.

**Pull request merged**